README.md
# OctoTuya-SmartPlug
With this plugin you'll be able to control [Tuya-based](https://en.tuya.com/) SmartPlugs either directly from Octoprint Web interface or through GCODE commands<br>

Work based on [OctoPrint-TuyaSmartplug](https://github.com/ziirish/OctoPrint-TuyaSmartplug). <br>
The original plugin by [ziirish](https://github.com/ziirish) was a must have to my 3D printing setup since my printer is connect to a Tuya smart outlet.<br>
But after going to the whole process of finding the required information to use the plugin I found out that it was not working properly and its been a couple of months since it was not updated, therefore forking the original repository and opening a Pull Request wasn't the best option.<br>
So I decide to look into its code and work the problems that I found, adding some changes to improve his work and a nice Wiki to guide users to the process of using it.<br>
This version of the plugin uses the tinytuya API to do the communications instead of the pytuya API. The tinytuya proved in my tests to be more reliable in terms of keeping comunication stable with the device and be more compatible as well.

<br>

## Disclaimer

Tuya is by far the most difficult IoT plataform that can be used to control devices by third-part software like this plugin, they require very specific information on the devices and change often their IoT Cloud Service.<br>
Many user have given up on using Tuya devices with OctoPrint because of its difficulty, some changed to other brands, other dug deeper and changed the device firmware or completely changed the device microcontroler, literally soldering a new one.<br>
If you, like me, only have Tuya devices and doesn't feel confortable on doing firmware flashes or resoldering, this plugin has a place on your OctoPrint instalation and its worth of your time configuring it. <br><br>
This plugin still needs improvements on the code itself, UI, performance and other elements, so you are more then welcome to open [PullRequests](https://github.com/andrelucca/OctoTuya-SmartPlug/pulls) with code updates/fixes or [Issues](https://github.com/andrelucca/OctoTuya-SmartPlug/issues) regarding what needs to be fixed. <br>
This is of course a side project of mine that unites two subjects that I like, so if you open a PR, Issue or a Discussion topic I'll answer as soon as I can, don't be angry if it takes more time than you think it should :). 

## How it was tested?

I tested all the plugin features using my Ender 3 V2 (Using original Marlin as Firmware that I configured and compiled) connected to the Octoprint v1.8.7. The Raspberry of my Octoprint is connected on the PSU of the printer, so if I power of the printer using the Tuya outlet it will also power-off the Raspberry (and will do it unsafely if I hadn't shutdown the Pi OS properly).

## Setup

Install via the bundled [Plugin Manager](https://github.com/foosel/OctoPrint/wiki/Plugin:-Plugin-Manager)
or manually using this URL:

    https://github.com/andrelucca/OctoTuya-SmartPlug/archive/main.zip

## Preparatory Work

All Tuya devices requires 2 infos in order to be controled by other software: `Device ID` and `Local Key`<br>
And this is were the greater difficulty takes place. I made a full guide on how obtain this info in this project [Wiki](https://github.com/andrelucca/OctoTuya-SmartPlug/wiki) so make sure you follow it and have this infos before installing an using this plugin.

## Configuration and Settings

All details on how to configure the plugin and how to change its settings according to your liking are in this project [Wiki](https://github.com/andrelucca/OctoTuya-SmartPlug/wiki)

## Support jneilliii and ziirish Efforts
Most of the code used in this plugin has been written by
[jneilliii](https://github.com/jneilliii) and its general idea came from [ziirish](https://github.com/ziirish) so if you want to support someone,
you can support their work.


changelog.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [0.1.0] - 2023-03-25
### Added
- Initial release based on [OctoPrint-octotuya by ziirish](https://github.com/ziirish/OctoPrint-octotuya).

[0.1.0]: https://github.com/andrelucca/OctoTuya-SmartPlug/tree/0.1.0


ultimaker_cura_script/TurnOffPrinter.py
# Copyright (c) 2017 Ghostkeeper
# The PostProcessingPlugin is released under the terms of the AGPLv3 or higher.

import re #To perform the search and replace.

from ..Script import Script


class TurnOffPrinter(Script):
    """Insert the Turn Off Action from Octotuya-SmartPlug Octoprint plugin 
    in the end of the gcode
    """

    def getSettingDataString(self):
        return """{
            "name": "Turn Off Printer",
            "key": "TurnOffPrinter",
            "metadata": {},
            "version": 2,
            "settings":
            {
                 "enabled":
                {
                    "label": "Enabled",
                    "description": "When enabled, the M81 command + the Label will be placed in the last line of the GCODE",
                    "type": "bool",
                    "default_value": false
                },
                "label":
                {
                    "label": "Label",
                    "description": "The configured label in the Octotuya-SmartPlug Octoprint plugin",
                    "type": "str",
                    "default_value": ""
                },
                "command": {
                    "label": "Command",
                    "description": "GCODE command to turn off the printer",
                    "type": "enum",
                    "options": {
                        "M81": "M81",
                        "G4 P2": "G4 P2"
                    },
                    "default_value": "M81"
                }
            }
        }"""

    def execute(self, data):
        label_string = self.getSettingValueByKey("label")
        command = self.getSettingValueByKey("command")
        if self.getSettingValueByKey("enabled"):
            
            seach_string_escape = re.escape(";End of Gcode")
            seach_string_compile = re.compile(seach_string_escape)
            turn_off_string = command+' '+label_string+' ;Turn off printer using OctoTuya\n;End of Gcode'

            for layer_number, layer in enumerate(data):
                data[layer_number] = re.sub(seach_string_compile, turn_off_string, layer)
            return data
        else:
            return data

octotuya_smartplug/static/js/octotuya.js
/*
 * View model for OctoTuya-SmartPlug
 *
 * Author: ziirish
 * License: AGPLv3
 */
$(function() {
    function octotuyaViewModel(parameters) {
        var self = this;

        self.settings = parameters[0];
		self.loginState = parameters[1];

		self.arrSmartplugs = ko.observableArray();
		self.isPrinting = ko.observable(false);
		self.selectedPlug = ko.observable();
		self.processing = ko.observableArray([]);

		self.onBeforeBinding = function() {
			self.arrSmartplugs(self.settings.settings.plugins.octotuya.arrSmartplugs());
        }

		self.onAfterBinding = function() {
			self.checkStatuses();
		}

        self.onEventSettingsUpdated = function(payload) {
			self.arrSmartplugs(self.settings.settings.plugins.octotuya.arrSmartplugs());
		}

		self.onEventPrinterStateChanged = function(payload) {
			if (payload.state_id == "PRINTING" || payload.state_id == "PAUSED"){
				self.isPrinting(true);
			} else {
				self.isPrinting(false);
			}
		}

		self.cancelClick = function(data) {
			self.processing.remove(data.label());
		}

		self.editPlug = function(data) {
			self.selectedPlug(data);
			$("#TuyaPlugEditor").modal("show");
		}

		self.addPlug = function() {
			self.selectedPlug({'ip':ko.observable(''),
									'id':ko.observable(''),
									'slot':ko.observable(1),
									'localKey':ko.observable(''),
									'label':ko.observable(''),
									'icon':ko.observable('icon-bolt'),
									'displayWarning':ko.observable(true),
									'v33':ko.observable(false),
									'warnPrinting':ko.observable(false),
									'gcodeEnabled':ko.observable(false),
									'gcodeOnDelay':ko.observable(0),
									'gcodeOffDelay':ko.observable(0),
									'autoConnect':ko.observable(true),
									'autoConnectDelay':ko.observable(10.0),
									'autoDisconnect':ko.observable(true),
									'autoDisconnectDelay':ko.observable(0),
									'sysCmdOn':ko.observable(false),
									'sysRunCmdOn':ko.observable(''),
									'sysCmdOnDelay':ko.observable(0),
									'sysCmdOff':ko.observable(false),
									'sysRunCmdOff':ko.observable(''),
									'sysCmdOffDelay':ko.observable(0),
									'currentState':ko.observable('unknown'),
									'btnColor':ko.observable('#808080'),
									'useCountdownRules':ko.observable(false),
									'countdownOnDelay':ko.observable(0),
									'countdownOffDelay':ko.observable(0)});
			self.settings.settings.plugins.octotuya.arrSmartplugs.push(self.selectedPlug());
			$("#TuyaPlugEditor").modal("show");
		}

		self.removePlug = function(row) {
			self.settings.settings.plugins.octotuya.arrSmartplugs.remove(row);
		}

		self.onDataUpdaterPluginMessage = function(plugin, data) {
            if (plugin != "octotuya") {
                return;
            }

			plug = ko.utils.arrayFirst(self.settings.settings.plugins.octotuya.arrSmartplugs(),function(item){
				return item.label() === data.label;
				}) || {'label':data.label,'currentState':'unknown','btnColor':'#808080'};

			if (plug.currentState != data.currentState) {
				plug.currentState(data.currentState)
				switch(data.currentState) {
					case "on":
						break;
					case "off":
						break;
					default:
						new PNotify({
							title: 'Tuya Smartplug Error',
							text: 'Status ' + plug.currentState() + ' for ' + plug.ip() + '. Double check IP Address\\Device ID\\Local Key in OctoTuya-SmartPlug Settings.',
							type: 'error',
							hide: true
							});
				self.settings.saveData();
				}
			}
			self.processing.remove(data.label);
        };

		self.toggleRelay = function(data) {
			self.processing.push(data.label());
			switch(data.currentState()){
				case "on":
					self.turnOff(data);
					break;
				case "off":
					self.turnOn(data);
					break;
				default:
					self.checkStatus(data.label());
			}
		}

		self.turnOn = function(data) {
/* 			if(data.sysCmdOn()){
				setTimeout(function(){self.sysCommand(data.sysRunCmdOn())},data.sysCmdOnDelay()*1000);
			} */
			self.sendTurnOn(data);
		}

		self.sendTurnOn = function(data) {
            $.ajax({
                url: API_BASEURL + "plugin/octotuya",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "turnOn",
					label: data.label()
                }),
                contentType: "application/json; charset=UTF-8"
            });
        };

    	self.turnOff = function(data) {
			if((data.displayWarning() || (self.isPrinting() && data.warnPrinting())) && !$("#octotuyaWarning").is(':visible')){
				self.selectedPlug(data);
				$("#octotuyaWarning").modal("show");
			} else {
				$("#octotuyaWarning").modal("hide");
/* 				if(data.sysCmdOff()){
					setTimeout(function(){self.sysCommand(data.sysRunCmdOff())},data.sysCmdOffDelay()*1000);
				} */
				self.sendTurnOff(data);
			}
        };

		self.sendTurnOff = function(data) {
			$.ajax({
			url: API_BASEURL + "plugin/octotuya",
			type: "POST",
			dataType: "json",
			data: JSON.stringify({
				command: "turnOff",
				label: data.label()
			}),
			contentType: "application/json; charset=UTF-8"
			});
		}

		self.checkStatus = function(plugLabel) {
            $.ajax({
                url: API_BASEURL + "plugin/octotuya",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "checkStatus",
					label: plugLabel
                }),
                contentType: "application/json; charset=UTF-8"
            }).done(function(){
				self.settings.saveData();
				});
        };

		self.disconnectPrinter = function() {
            $.ajax({
                url: API_BASEURL + "plugin/octotuya",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "disconnectPrinter"
                }),
                contentType: "application/json; charset=UTF-8"
            });
		}

		self.connectPrinter = function() {
            $.ajax({
                url: API_BASEURL + "plugin/octotuya",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "connectPrinter"
                }),
                contentType: "application/json; charset=UTF-8"
            });
		}

/* 		self.sysCommand = function(sysCmd) {
            $.ajax({
                url: API_BASEURL + "plugin/octotuya",
                type: "POST",
                dataType: "json",
                data: JSON.stringify({
                    command: "sysCommand",
					cmd: sysCmd
                }),
                contentType: "application/json; charset=UTF-8"
            });
		} */

		self.checkStatuses = function() {
			ko.utils.arrayForEach(self.settings.settings.plugins.octotuya.arrSmartplugs(),function(item){
				if(item.label() !== "") {
					console.log("checking " + item.label())
					self.checkStatus(item.label());
				}
			});
			if (self.settings.settings.plugins.octotuya.pollingEnabled()) {
				setTimeout(function() {self.checkStatuses();}, (parseInt(self.settings.settings.plugins.octotuya.pollingInterval(),10) * 60000));
			};
        };
    }

    // view model class, parameters for constructor, container to bind to
    OCTOPRINT_VIEWMODELS.push([
        octotuyaViewModel,

        // e.g. loginStateViewModel, settingsViewModel, ...
        ["settingsViewModel","loginStateViewModel"],

        // "#navbar_plugin_octotuya","#settings_plugin_octotuya"
        ["#navbar_plugin_octotuya","#settings_plugin_octotuya"]
    ]);
});


octotuya_smartplug/__init__.py
# coding=utf-8
from __future__ import absolute_import

import octoprint.plugin
from octoprint.server import user_permission
import socket
import json
import logging
import os
import re
import threading
import time
import tinytuya


class octotuyaPlugin(
    octoprint.plugin.SettingsPlugin,
    octoprint.plugin.AssetPlugin,
    octoprint.plugin.TemplatePlugin,
    octoprint.plugin.SimpleApiPlugin,
    octoprint.plugin.StartupPlugin,
):
    def __init__(self):
        self._logger = logging.getLogger("octoprint.plugins.octotuya")
        self._octotuya_logger = logging.getLogger(
            "octoprint.plugins.octotuya.debug"
        )

    # ~~ StartupPlugin mixin

    def on_startup(self, host, port):
        # setup customized logger
        from octoprint.logging.handlers import CleaningTimedRotatingFileHandler

        octotuya_logging_handler = CleaningTimedRotatingFileHandler(
            self._settings.get_plugin_logfile_path(postfix="debug"),
            when="D",
            backupCount=3,
        )
        octotuya_logging_handler.setFormatter(
            logging.Formatter("[%(asctime)s] %(levelname)s: %(message)s")
        )
        octotuya_logging_handler.setLevel(logging.DEBUG)

        self._octotuya_logger.addHandler(octotuya_logging_handler)
        self._octotuya_logger.setLevel(
            logging.DEBUG
            if self._settings.get_boolean(["debug_logging"])
            else logging.INFO
        )
        self._octotuya_logger.propagate = False

    def on_after_startup(self):
        self._logger.info("octotuya loaded!")

    # ~~ SettingsPlugin mixin

    def get_settings_defaults(self):
        return dict(
            debug_logging=False,
            arrSmartplugs=[
                {
                    "ip": "",
                    "id": "",
                    "slot": 1,
                    "localKey": "",
                    "label": "",
                    "icon": "icon-bolt",
                    "displayWarning": True,
                    "warnPrinting": False,
                    "gcodeEnabled": False,
                    "v33": False,
                    "gcodeOnDelay": 0,
                    "gcodeOffDelay": 0,
                    "autoConnect": True,
                    "autoConnectDelay": 10.0,
                    "autoDisconnect": True,
                    "autoDisconnectDelay": 0,
                    "sysCmdOn": False,
                    "sysRunCmdOn": "",
                    "sysCmdOnDelay": 0,
                    "sysCmdOff": False,
                    "sysRunCmdOff": "",
                    "sysCmdOffDelay": 0,
                    "currentState": "unknown",
                    "btnColor": "#808080",
                    "useCountdownRules": False,
                    "countdownOnDelay": 0,
                    "countdownOffDelay": 0,
                }
            ],
            pollingInterval=15,
            pollingEnabled=False,
        )

    def get_settings_restricted_paths(self):
        return dict(
            admin=[
                [
                    "arrSmartplugs",
                ],
            ]
        )

    def on_settings_save(self, data):
        old_debug_logging = self._settings.get_boolean(["debug_logging"])

        octoprint.plugin.SettingsPlugin.on_settings_save(self, data)

        new_debug_logging = self._settings.get_boolean(["debug_logging"])
        if old_debug_logging != new_debug_logging:
            if new_debug_logging:
                self._octotuya_logger.setLevel(logging.DEBUG)
            else:
                self._octotuya_logger.setLevel(logging.INFO)

    def get_settings_version(self):
        return 3

    def on_settings_migrate(self, target, current=None):
        if current is None or current < self.get_settings_version():
            # Reset plug settings to defaults.
            self._logger.debug("Resetting arrSmartplugs for octotuya settings.")
            self._settings.set(
                ["arrSmartplugs"], self.get_settings_defaults()["arrSmartplugs"]
            )

    # ~~ AssetPlugin mixin

    def get_assets(self):
        return dict(js=["js/octotuya.js"], css=["css/octotuya.css"])

    # ~~ TemplatePlugin mixin

    def get_template_configs(self):
        return [
            dict(type="navbar", custom_bindings=True),
            dict(type="settings", custom_bindings=True),
        ]

    # ~~ SimpleApiPlugin mixin

    def turn_on(self, pluglabel):
        self._octotuya_logger.debug("Turning on %s." % pluglabel)
        if self.is_turned_on(pluglabel=pluglabel):
            self._octotuya_logger.debug("Plug %s already turned on" % pluglabel)
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="on", label=pluglabel)
            )
            return
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        self._octotuya_logger.debug(plug)
        if plug["useCountdownRules"]:
            chk = self.sendCommand(
                "countdown", plug["label"], int(plug["countdownOnDelay"])
            )
        else:
            chk = self.sendCommand("on", plug["label"])

        if chk is not False:
            self.check_status(plug["label"], chk)
            if plug["autoConnect"]:
                c = threading.Timer(
                    int(plug["autoConnectDelay"]), self._printer.connect
                )
                c.start()
                if plug["sysCmdOn"]:
                    t = threading.Timer(
                        int(plug["sysCmdOnDelay"]),
                        os.system,
                        args=[plug["sysRunCmdOn"]],
                    )
                    t.start()
                else:
                    self._plugin_manager.send_plugin_message(
                        self._identifier, dict(currentState="unknown", label=pluglabel)
                    )

    def turn_off(self, pluglabel):
        self._octotuya_logger.debug("Turning off %s." % pluglabel)
        if not self.is_turned_on(pluglabel=pluglabel):
            self._octotuya_logger.debug("Plug %s already turned off" % pluglabel)
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="off", label=pluglabel)
            )
            return
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        self._octotuya_logger.debug(plug)
        if plug["useCountdownRules"]:
            chk = self.sendCommand(
                "countdown", plug["label"], int(plug["countdownOffDelay"])
            )

        if plug["sysCmdOff"]:
            t = threading.Timer(
                int(plug["sysCmdOffDelay"]), os.system, args=[plug["sysRunCmdOff"]]
            )
            t.start()
            if plug["autoDisconnect"]:
                self._printer.disconnect()
                time.sleep(int(plug["autoDisconnectDelay"]))

        if not plug["useCountdownRules"]:
            chk = self.sendCommand("off", plug["label"])

        if chk is not False:
            self.check_status(plug["label"], chk)
        else:
            self._plugin_manager.send_plugin_message(
                self._identifier, dict(currentState="unknown", label=pluglabel)
            )

    def check_status(self, pluglabel, resp=None):
        self._octotuya_logger.debug("Checking status of %s." % pluglabel)
        if pluglabel != "":
            response = resp or self.sendCommand("info", pluglabel)
            if not isinstance(response, dict) or "Error" in response:
                self._octotuya_logger.warning(
                    "Unable to check device status: %s" % response
                )
                self._plugin_manager.send_plugin_message(
                    self._identifier, dict(currentState="unknown", label=pluglabel)
                )
            else:
                self._plugin_manager.send_plugin_message(
                    self._identifier,
                    dict(
                        currentState=(
                            "on" if self.is_turned_on(response, pluglabel) else "off"
                        ),
                        label=pluglabel,
                    ),
                )

    def is_turned_on(self, data=None, pluglabel=None):
        if data is None and pluglabel:
            data = self.sendCommand("info", pluglabel)

        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        return data and plug and data.get("dps", {}).get(str(plug["slot"]))

    def get_api_commands(self):
        return dict(turnOn=["label"], turnOff=["label"], checkStatus=["label"])

    def on_api_command(self, command, data):
        if not user_permission.can():
            from flask import make_response

            return make_response("Insufficient rights", 403)

        if command == "turnOn":
            self.turn_on("{label}".format(**data))
        elif command == "turnOff":
            self.turn_off("{label}".format(**data))
        elif command == "checkStatus":
            self.check_status("{label}".format(**data))

    # ~~ Utilities

    def plug_search(self, lst, key, value):
        for item in lst:
            if item[key] == value:
                return item

    def sendCommand(self, cmd, pluglabel, args=None):
        self._octotuya_logger.debug("Sending command: %s to %s" % (cmd, pluglabel))
        plug = self.plug_search(
            self._settings.get(["arrSmartplugs"]), "label", pluglabel
        )
        device = tinytuya.OutletDevice(plug["id"], plug["ip"], plug["localKey"])
        if plug.get("v33"):
            device.set_version(3.3)

        commands = {
            "info": ("status", None),
            "on": ("set_status", True),
            "off": ("set_status", False),
            "countdown": ("set_timer", None),
        }

        command, arg = commands[cmd]
        func = getattr(device, command, None)
        if not func:
            self._octotuya_logger.debug("No such command '%s'" % command)
            return False
        if args:
            func(args)
        elif arg is not None:
            func(arg, plug["slot"])
        else:
            func()
            time.sleep(0.5)
            ret = device.status()
            self._octotuya_logger.debug("Status: %s" % str(ret))
            return ret

    # ~~ Gcode processing hook

    def gcode_turn_off(self, plug):
        if plug["warnPrinting"] and self._printer.is_printing():
            self._logger.info(
                "Not powering off %s because printer is printing." % plug["label"]
            )
        else:
            self.turn_off(plug["label"])

    def processGCODE(self, comm_instance, phase, cmd, cmd_type, gcode, *args, **kwargs):
        if gcode:
            if cmd.startswith("M80"):
                name = re.sub(r"^M80\s?", "", cmd)
                self._octotuya_logger.debug(
                    "Received M80 command, attempting power on outlet of label %s." % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "label", name
                )
                if not plug:
                    self._octotuya_logger.debug("No outlet found with label %s." % name)
                else:  
                    self._octotuya_logger.debug(plug)
                    if plug["gcodeEnabled"]:
                        t = threading.Timer(
                            int(plug["gcodeOnDelay"]),
                            self.turn_on,
                            args=[plug["label"]],
                        )
                        t.start()
                        return
                    else:
                        return
            elif cmd.startswith("M81"):
                name = re.sub(r"^M81\s?", "", cmd)
                self._octotuya_logger.debug(
                    "Received M81 command, attempting power off outlet of label %s." % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "label", name
                )
                if not plug:
                    self._octotuya_logger.debug("No outlet found with label %s." % name)
                else:  
                    self._octotuya_logger.debug(plug)
                    if plug["gcodeEnabled"]:
                        t = threading.Timer(
                            int(plug["gcodeOffDelay"]),
                            self.gcode_turn_off,
                            args=[plug],
                        )
                        t.start()
                        return
                    else:
                        return
            elif cmd.startswith("G4 P1"):
                name = re.sub(r"^G4 P1\s?", "", cmd)
                self._octotuya_logger.debug(
                    "Received @TUYAON command, attempting power on outlet of label %s."
                    % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]), "label", name
                )
                if not plug:
                    self._octotuya_logger.debug("No outlet found with label %s." % name)
                else:  
                    self._octotuya_logger.debug(plug)
                    if plug["gcodeEnabled"]:
                        t = threading.Timer(
                            int(plug["gcodeOnDelay"]),
                            self.turn_on,
                            args=[plug["label"]],
                        )
                        t.start()
                        return
                    else:
                        return
            elif cmd.startswith("G4 P2"):
                name = re.sub(r"^G4 P2\s?", "", cmd)
                self._octotuya_logger.debug(
                    "Received @TUYAOFF command, attempting power off outlet of label %s."
                    % name
                )
                plug = self.plug_search(
                    self._settings.get(["arrSmartplugs"]),
                    "label",
                    name,
                )
                if not plug:
                    self._octotuya_logger.debug("No outlet found with label %s." % name)
                else:    
                    self._octotuya_logger.debug(plug)
                    if plug["gcodeEnabled"]:
                        t = threading.Timer(
                            int(plug["gcodeOffDelay"]),
                            self.gcode_turn_off,
                            args=[plug],
                        )
                        t.start()
                        return
                    else:
                        return

    # ~~ Softwareupdate hook

    def get_update_information(self):
        # Define the configuration for your plugin to use with the Software Update
        # Plugin here. See https://github.com/foosel/OctoPrint/wiki/Plugin:-Software-Update
        # for details.
        return dict(
            octotuya=dict(
                displayName="OctoTuya SmartPlug",
                displayVersion=self._plugin_version,
                # version check: github repository
                type="github_release",
                user="andrelucca",
                repo="OctoTuya-SmartPlug",
                current=self._plugin_version,
                # update method: pip
                pip="https://github.com/andrelucca/OctoTuya-SmartPlug/archive/{target_version}.zip",
            )
        )


# If you want your plugin to be registered within OctoPrint under a different name than what you defined in setup.py
# ("OctoPrint-PluginSkeleton"), you may define that here. Same goes for the other metadata derived from setup.py that
# can be overwritten via __plugin_xyz__ control properties. See the documentation for that.
__plugin_name__ = "OctoTuya SmartPlug"
__plugin_version__ = "0.1.0"
__plugin_pythoncompat__ = ">=2.7,<4"


def __plugin_load__():
    global __plugin_implementation__
    __plugin_implementation__ = octotuyaPlugin()

    global __plugin_hooks__
    __plugin_hooks__ = {
        "octoprint.comm.protocol.gcode.queuing": __plugin_implementation__.processGCODE,
        "octoprint.plugin.softwareupdate.check_config": __plugin_implementation__.get_update_information,
    }
